UDP Conn implemented

nickaxgit opened the pull request:
Hi

Sorry - some of my comments are a bit sharp - I was a bit frustrated at the time - please don't be offended !

I have source for a simple demo (echo client .. ) if that's useful

usage is fundamentally like this:-

` conn.OpenDialUDP(1042, routerMAC, svAddr)
	 
	 buff := make([]byte,64)

	 for{	
		println("Writing")
	 	conn.Write([]byte("hello"))
	 	
		time.Sleep(1* time.Second)

		println("reading")
	 	n,err  := conn.Read(buff)
	 	if err != io.EOF{
	 		println("Read:",n, " bytes",string(buff))
			break
	 	}
		
	 }

	 conn.Close()
`

Not absolutely sure what to do with some of the things that don't really apply to UDP - setWriteDeadline, close etc ... but what is there is working well for my purposes - I built a very simple sequenced numbered, retrying "reliable" protocol on top (a watered down TCP) and able to receive snapshots from my remote camera on a cellular connection - it's working great .

Cheers

Nick

soypat commented:
Thank you a bunch for this work nick! Will try to get around to looking at it this weekend, if not sometime throughout the week. Still settling down after a long week at Gophercon :)

nickaxgit commented:
Sure, no rush

I hope you had a good one

I *think* I may have found a ‘bug’ in the bmp280 driver (temperature and
humidity sensor) not sure if that’s relevant to you .. it involves clearing
some reserved bits in a register, which then causes spurious pressure
readings..

And I have the bones of a ‘driver’ for  OV2640 based SPI cameras (and it’s
not a big leap to support the full ArduCam range)

I just don’t know how much time I can commit . . . I can see how it would
be easy to get very sucked in, but I need to make a living at something.

I can certainly try and contribute, wherever my own goals align.

I love that TinyGo Creates native executables, and that (I presume) they
massively outperform Python, and come close to C - I also want/need PIO
support, and I think we have that right ?

Cheers

Nick



On Sat, 13 Jul 2024 at 17:01, Patricio Whittingslow <
***@***.***> wrote:

> Thank you a bunch for this work nick! Will try to get around to looking at
> it this weekend, if not sometime throughout the week. Still settling down
> after a long week at Gophercon :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/soypat/seqs/pull/29#issuecomment-2226973283>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALR2VA5SBPFZOFRTOY547NDZMFFL3AVCNFSM6AAAAABKY55NECVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMRWHE3TGMRYGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

soypat commented on eth/headers.go:
```diff
@@ -227,7 +227,8 @@ func (ehdr EthernetHeader) AssertType() EtherType { return EtherType(ehdr.SizeOr
 // Put marshals the ethernet frame onto buf. buf needs to be 14 bytes in length or Put panics.
 func (ehdr *EthernetHeader) Put(buf []byte) {
-	_ = buf[13]
+	_ = buf[13] 
```
Please discard these changes, I don't see a diff.

nickaxgit replied:
My fault - extra whitespace, I had commented line 230 - explaining the bounds check - but left a trailing space when removing it

soypat commented on stacks/port_tcp.go:
```diff
@@ -9,13 +9,23 @@ import (
+// </unhelpful and misleading !>
+
+// <suggeted replacement>
+// tcphandler represents a user provided function for handling incoming TCP packets on a port.
+// Incoming data is passed in a 'pkt' to the recv function which is invoked whenever data arrives (by RecvEth)
+// Outgoing data is written into the `dst` byte slice (from the tx ring buffer). The function must return the number of
+// bytes written to `response` and an error.
+// TCPConn provides an implemntation of this interface - note .send is ONLY called by HandleEth
```
No need for xml tags, github PRs show previous and new changes. Your changes reflect the latest API, so should be more helpful than a description of a previous API.

soypat commented on stacks/port_tcp.go:
```diff
@@ -9,13 +9,23 @@ import (
+
 // See [PortStack] for information on how to use this function and other port handlers.
-type itcphandler interface {
+type itcphandler interface {  //note TCPConn is our implementation of this interface 
```
good note. Include it in itcphandler documentation by putting it above the interface declaration.

soypat commented on stacks/port_tcp.go:
```diff
@@ -193,6 +203,45 @@ func (pkt *TCPPacket) CalculateHeaders(seg seqs.Segment, payload []byte) {
 	pkt.TCP.Checksum = pkt.TCP.CalculateChecksumIPv4(&pkt.IP, nil, payload)
 }
 
+func (pkt *UDPPacket) CalculateHeaders( payload []byte) {
```
clean the function body comments please

soypat commented on stacks/port_udp.go:
```diff
@@ -16,7 +16,7 @@ type iudphandler interface {
 
 type udpPort struct {
-	ihandler iudphandler
+	handler iudphandler
```
good eye. we don't want C# code smell hehe

soypat commented on stacks/port_udp.go:
```diff
@@ -25,32 +25,36 @@ func (port udpPort) Port() uint16 { return port.port }
 func (port *udpPort) HandleEth(dst []byte) (int, error) {
-	if port.ihandler == nil {
+	
+	//println("udpPort.HandleEth: ", string(dst)) //@@@REMOVE
```
@@@ Remove!

soypat commented on stacks/port_udp.go:
```diff
@@ -25,32 +25,36 @@ func (port udpPort) Port() uint16 { return port.port }
 }
 
-// Open sets the UDP handler and opens the port.
+// Open sets the UDP handler and opens the port. 
```
Why are unchanged lines included in the diff. Are you formatting your Go code with gofmt? This should not appear in a Go diff since gofmt is pretty good at discarding whitespace noise. It worries me that there is some gomt-removable whitespace included on all your lines changed which would make the next commit *very* noisy, which we don't want.

nickaxgit replied:
I have disabled an additional 3rd party linter that may have been responsible - i am using vscode and have the go extension installed

soypat commented on stacks/portstack.go:
```diff
@@ -167,7 +167,7 @@ func (ps *PortStack) RecvEth(ethernetFrame []byte) (err error) {
 		return errPacketSmol
 	} else if len(payload) > int(ps.mtu) {
-		println("recv", payload, ps.mtu)
+		//println("recv", payload, ps.mtu)
```
Better to remove altogether if modifying

nickaxgit replied:
global searched for println's all out

soypat commented on stacks/portstack.go:
```diff
@@ -376,29 +376,41 @@ func (ps *PortStack) HandleEth(dst []byte) (n int, err error) {
 //
 // If a handler returns any other error the port is closed.
 func (ps *PortStack) handleEth(dst []byte) (n int, err error) {
+	//this would be much better named as getNextOutboundPacket
+
+	//println ("portStack:handleEth ",string(dst)) //@@@REMOVE
+
```
This discussion belongs in github issue tracker! also remove the comment

nickaxgit replied:
Opened issue .. I do think it needs a better name - it was one area that was very opaque to me for a long time, the fact the slice is passed by refence (for population) is not super obvious when coming from other languages

soypat commented on stacks/portstack.go:
```diff
@@ -376,29 +376,41 @@ func (ps *PortStack) HandleEth(dst []byte) (n int, err error) {
+
 	switch {
 	case len(dst) < int(ps.mtu):
+		
```
Leave whitespace as is here and below please

soypat commented on stacks/portstack.go:
```diff
@@ -376,29 +376,41 @@ func (ps *PortStack) HandleEth(dst []byte) (n int, err error) {
+		 HandleEth(dst []byte) (int, error)
 	}
-
+	//or why this is an anonymous function is stored in a local variable before invocation  - I can see it narrows scope, but it's an odd pattern and confusing as the code does not run in the order it appears
```
We can extract both Socket and handleSocket to package level in this or a future PR. 

Note: Socket would have to be unexported, so `socket`,`isocket` or something.

nickaxgit replied:
done .. Socket is now socket, and handleSocket is now package level - will test before next PR

soypat commented on stacks/portstack.go:
```diff
@@ -376,29 +376,41 @@ func (ps *PortStack) HandleEth(dst []byte) (n int, err error) {
-
+	//or why this is an anonymous function is stored in a local variable before invocation  - I can see it narrows scope, but it's an odd pattern and confusing as the code does not run in the order it appears
 	handleSocket := func(dst []byte, sock Socket) (int, bool, error) {
+		
+		//println ("handleSocket called") //@@@REMOVE
+		//note sock is a UDPport or TCPport - things that impliment the Sock interface
+		
```
remove comments, the second comment is OK but belongs above the Socket type definition.

soypat commented on stacks/portstack.go:
```diff
@@ -376,29 +376,41 @@ func (ps *PortStack) HandleEth(dst []byte) (n int, err error) {
 			return 0, false, nil // Nothing to handle, just skip.
 		}
 		// Socket has an unhandled packet.
+		//println ("handleSocket, calling sock.handleEth") //@@@REMOVE
```
remove!

soypat commented on stacks/portstack.go:
```diff
@@ -419,6 +431,7 @@ func (ps *PortStack) handleEth(dst []byte) (n int, err error) {
 	isDebug := ps.isLogEnabled(slog.LevelDebug)
 	socketPending := false
 	if ps.pendingUDPv4 > 0 {
+		
```
Remove all whitepsace noise!

nickaxgit replied:
I can't get format on save to work, but doing it manually with alt-shift-f does work - super annoying

soypat replied:
This is my VSCode JSON file. Find it by going to settings and clicking the page icon top right corner
```
{
    "workbench.colorTheme": "GitHub Dark High Contrast",
    "workbench.sideBar.location": "right",
    "[go]": {
        "editor.insertSpaces": false,
        "editor.formatOnSave": true,
        "editor.codeActionsOnSave": {
            "source.organizeImports": "explicit"
        }
    },
    "gopls": {
        "ui.semanticTokens": true
    },
    "go.toolsManagement.autoUpdate": true
}
```

soypat commented on stacks/portstack.go:
```diff
@@ -427,8 +440,9 @@ func (ps *PortStack) handleEth(dst []byte) (n int, err error) {
+		
 				if isDebug {
-					ps.debug("UDP:send", slog.Int("plen", n))
+					ps.debug("UDP:send", slog.Int("plen", n))					
```
This definetly should have been caught by gofmt, please make sure your editor runs gofmt on Go file save. VSCode does this by default.

nickaxgit replied:
or not .. despite being set for both the user and the workspace, go selected as the formatter , and restarting vscode

<img width="1109" alt="image" src="https://github.com/user-attachments/assets/cd30154b-8651-439d-9992-a897bd470dec">

Manual formatting with alt,shift F works

soypat replied:
Hmm, I don't have that enabled. It seems Go formatting is independent of that option. See my JSON comment.

nickaxgit replied:
got you

soypat commented on stacks/portstack.go:
```diff
@@ -462,6 +476,7 @@ func (ps *PortStack) handleEth(dst []byte) (n int, err error) {
 	return 0, nil // Nothing handled.
 }
 
+
```
more noise that should be removed.

soypat commented on stacks/ring.go:
```diff
@@ -14,7 +14,11 @@ type ring struct {
 
+//writes b[] into the ring buffer at the current write position (r.end)  - wrapping if needed, and increasing r.end (possibly wrapping)
 func (r *ring) Write(b []byte) (int, error) {
+
+	//println("wrote ",string(b),"into tx ring buff") //@@@REMOVE
+	
```
remove whitespace and comment

soypat commented on stacks/ring.go:
```diff
@@ -14,7 +14,11 @@ type ring struct {
 	end int
 }
 
+//writes b[] into the ring buffer at the current write position (r.end)  - wrapping if needed, and increasing r.end (possibly wrapping)
```
Always start function docstrings with the function name, preceded by a space, so: `// Write ...`

soypat commented on stacks/ring.go:
```diff
@@ -105,11 +109,12 @@ func (r *ring) onReadEnd() {
 	}
 	if r.off == r.end {
-		r.Reset() // We read everything, reset.
+		r.Reset() // We read everything, reset.  - why reset - ever ?
```
Reduces split buffer reads, which invokes two memcpy instead of only one.

nickaxgit replied:
Ahh !

soypat commented on stacks/tcpconn.go:
```diff
@@ -39,7 +40,7 @@ type TCPConn struct {
 	remote    netip.AddrPort
 	localPort uint16
-	remoteMAC [6]byte
+	remoteMAC [6]byte //this is the ROUTERS mac address .. it's not a great name (IMHO) - unless it isn't in which case I appologise
```
It's not necessarily the router MAC, could be a peer on the same network

soypat commented on stacks/tcpconn.go:
```diff
@@ -50,12 +51,12 @@ type TCPConn struct {
 }
 
-type TCPConnConfig struct {
+type TCPConConfig struct {
```
Woah, revert back.

nickaxgit replied:
Sorry - done (not sure how that happened)

soypat commented on stacks/tcpconn.go:
```diff
@@ -160,8 +161,10 @@ func (sock *TCPConn) Write(b []byte) (n int, _ error) {
+
+// Read reads data from the socket's input (RX) (ring) buffer... populating b[].. 
+//If the rx buffer is empty, Read will block until data is available.
+
```
Please follow `//` with a space and make it adjacent to function name so it appears as part of documentation.

soypat commented on stacks/tcpconn.go:
```diff
@@ -331,6 +334,8 @@ func (sock *TCPConn) checkPipeOpen() error {
 
+//take a packet from the stack and and write it's payload into the socket's rx (ring) buffer
+// /\ i found this really misleading - the packet is not 'taken' from the stack, it is passed to the handler (called by PortStack RecvEth)
 func (sock *TCPConn) recv(pkt *TCPPacket) (err error) {
```
See function documenting methodology for Go, also feel free to provide documentation changes as part of the PR, no need to discuss it in comments, the discussion can happen in comments here. More easy to follow.

soypat commented on stacks/tcpconn.go:
```diff
@@ -380,6 +385,7 @@ func (sock *TCPConn) recv(pkt *TCPPacket) (err error) {
 	return err
 }
 
+//this handler is called by the underlying stack and populates response[] from the TX ring buffer, with data to be sent as a packet
```
Function documenting methodology

soypat commented on stacks/tcpconn.go:
```diff
@@ -414,14 +420,15 @@ func (sock *TCPConn) send(response []byte) (n int, err error) {
 		n, err = sock.tx.Read(payload)
 		if err != nil && err != io.EOF || n != int(seg.DATALEN) {
-			panic("bug in handleUser") // This is a bug in ring buffer or a race condition.
+			panic("bug in handleUser") // This is a bug in ring buffer or a race condition. - is it ? - odd message then
```
We can do with a better message, maybe something like: `unexpected condition in seqs.TCPConn.send`

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+	
+}
+
+func (u *UDPConn) SetReadDeadline(t time.Time) error  {
+    // not really relevant to UDP - buit the interface requires it
+	return nil
+}
+
+func (u *UDPConn) SetWriteDeadline(t time.Time) error  {
+    // not really relevant to UDP
+	return nil
+}
```
Return `errors.ErrUnsupported` from package `errors`.

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+
+
+func (u *UDPConn) Close() error {    
+    return nil
```
Close should close the connection as in PortStack should release the port for future use. As of now your Conn leaks after opening, it hogs the port forever.

nickaxgit replied:
found closeUDP (well, copilot did !)

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+//this handler is called regularly by the underlying stack (HandleEth) and populates response[] from the TX ring buffer, with data to be sent as a packet
+func (sock *UDPConn) send(response []byte) (n int, err error) {
+	
+	defer sock.trace("UDPConn.send:start")  //why is this deferred ? - coul dbe confusing
```
should not be deferred

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+//this is an identical structure to TCPConnConfig - but I didn't want to change the original name without discussions
+type BuffConfig struct { 
+	TxBufSize uint16
+	RxBufSize uint16
+}
```
UDPConnConfig

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+	// err := sock.scb.Open(iss, seqs.Size(len(sock.rx.buf)), state)
+	// if err != nil {
+	// 	return err
+	// }
```
remove tcp comments

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+package stacks
```
please clean up comments in this file. there seems to be lots of comments left over from a copy-pase operation.

soypat commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,230 @@
+	pkt.UDP.DestinationPort = sock.remote.Port()
+	pkt.Eth.Destination = sock.remoteMAC
+}
+
```
I'll give this file close look sometime throughout the week

b0ch3nski commented on stacks/ring.go:
```diff
@@ -14,7 +14,11 @@ type ring struct {
+// Write writes b[] into the ring buffer at the current write position (r.end)  - wrapping if needed, and increasing r.end (possibly wrapping)
 func (r *ring) Write(b []byte) (int, error) {
+
+	//println("wrote ",string(b),"into tx ring buff") //@@@REMOVES
```
Leftover?

nickaxgit replied:
yep - sorry - not sure on the workflow here - if i just fix and commit - does it create a whole lot of noise again ?

b0ch3nski replied:
Don't worry, GitHub is clever, you can go either way: make changes, create new commit and push it or make changes, squash & amend commits to single one and then push.

Some projects favor one over the other - personally I prefer the second approach as it keeps git history cleaner :slightly_smiling_face:

nickaxgit commented on stacks/udpconn.go:
```diff
@@ -0,0 +1,219 @@
+func (sock *UDPConn) Write(b []byte) (n int, err error) {
+	err = sock.stack.FlagPendingUDP(sock.localPort)
+	if err != nil {
+		panic("flagpending error")
```
yes .. I thought I'd copied the pattern from TCPConn .. but you are correct - it should return

soypat reviewed (changes requested):
Hey Nick! Some documentation changes are pending. I'd like to do some cleanup and add tests sometime this week before merging to main so I've added a new `dev` branch and rebased onto it.

nickaxgit commented:
Morning - My settings file .. I am missing a few of your go config options - should I include them ?

It's interesting my formatOnSave is false - I had been trying to change it through the command palette after a google search but could never get it to stick

{
  "[go]": {
    "editor.formatOnSave": false,
    "editor.codeActionsOnSave": {
      "source.organizeImports": "never"
    },
    "editor.defaultFormatter": "golang.go"
  },
  "go.formatTool": "gofmt",
  "settingsSync.ignoredExtensions": [],
  "notebook.kernelProviderAssociations": [],
  "editor.formatOnSave": true,
  "terminal.integrated.automationShell.windows": "",
  "git.autofetch": true,
  "html.autoClosingTags": false,
  "git.enableSmartCommit": true,
  "diffEditor.ignoreTrimWhitespace": false,
  "liveServer.settings.donotShowInfoMsg": true,
  "terminal.integrated.tabs.enabled": true,
  "terminal.integrated.defaultProfile.windows": "Command Prompt",
  "terminal.integrated.profiles.windows": {
    "C:\\Windows\\System32\\cmd.exe (migrated)": {
      "path": "C:\\Windows\\System32\\cmd.exe",
      "args": [
        "-ExecutionPolicy",
        "Bypass"
      ]
    }
  },
  "files.exclude": {
    // Hide all .js files when have a .ts file
    "**/*.js": {
      "when": "$(basename).ts"
    },
    "**/*.js.map": true
  },
  "workbench.editorAssociations": {
    "*.ipynb": "jupyter-notebook"
  },
  "notebook.cellToolbarLocation": {
    "default": "right",
    "jupyter-notebook": "left"
  },
  "[python]": {
    "editor.formatOnType": true
  },
  "cmake.configureEnvironment": {
    "PICO_SDK": "c:\\pico-sdk"
  },
  "cmake.configureOnOpen": true,
  "cmake.configureSettings": {},
  "editor.fontSize": 18,
  "window.zoomLevel": 1,
  "micropico.manualComDevice": "com9:",
  "cmake.pinnedCommands": [
    "workbench.action.tasks.configureTaskRunner",
    "workbench.action.tasks.runTask"
  ],
  "arduino.useArduinoCli": true,
  "editor.defaultFormatter": "tinygo.vscode-tinygo"
}

nickaxgit commented:
" so I've added a new `dev` branch and rebased onto it."

Not sure what the implications are for me.. do I keep committing/pushing to my fork - or do I need to rebase that (or something .. out of my depth here)

soypat commented:
@nickaxgit No special action needed on your part- just fix the issues and eventually your changes will be in main when I finish writing the tests.

soypat commented:
Nick! Thank you for your contribution! :smile: 

Merging into dev, discussion should continue in https://github.com/soypat/seqs/pull/31
